### PR TITLE
Remove unused STC.tls

### PR DESCRIPTION
**Pull request by Geod24**
```
STC.tls was originally set on __thread variables.
However, __thread was removed from the language in 2013
(20ef96fe1d5c1631f5aead01873f17ca85cc944b),
as TLS was made the default for global variables.
The STC was however left in place, but was only ever set
in one place: for the gate in [shared] static {c,d}tors.
However, this was likely to be an accident, and it also looks wrong:
the gate for shared static {c,d}tor was static (which is TLS),
while the gate for non-shared module {c,d}tors was STC.tls,
which had no effect (so it was effectively a local variable).

However, that bug was not visible, thanks to the glue layer,
which has special handling of gate variables, essentially
ignoring the STC.
```

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by Geod24** [comment]
```
../.dub/packages/eventcore-0.9.7/eventcore/source/eventcore/drivers/posix/processes.d: Error: variable `eventcore.drivers.posix.processes.PosixEventDriverProcesses!(EpollEventLoop).PosixEventDriverProcesses._sharedStaticCtor_L306_C2_1.__gate` __gshared not allowed in safe functions; use shared
```

Interesting, this is not covered by the test-suite. And the message could use improvement. I changed it to `shared`, let's see if the CI prefers this. Will raise another PR for this improvement.

**Review by WalterBright — APPROVED** [review]

**Review by UplinkCoder — APPROVED** [review]
Nice simplification.

**Comment by UplinkCoder** [comment]
Error: read-modify-write operations are not allowed for `shared` variables
       Use `core.atomic.atomicOp!"+="(__gate, 1)` instead
runnable/xtest46_gc.d-mixin-37(4699): Error: template instance
`xtest46_gc.Hoge6691!()` error instantiating


Am Do., 6. Aug. 2020 um 23:22 Uhr schrieb Walter Bright <
notifications@github.com>:

**Comment by Geod24** [comment]
I added a teeny tiny hack to work around the issue and documented in, plus created an issue for the larger bug, which is probably a good starter issue: https://issues.dlang.org/show_bug.cgi?id=21147

**Comment by MoonlightSentinel** [comment]
This should fix the incorrect module ctor behaviour in `runnable/test18868.d`:

```diff
diff --git a/src/dmd/dsymbolsem.d b/src/dmd/dsymbolsem.d
index a8d4b806e..3b2b8c575 100644
--- a/src/dmd/dsymbolsem.d
+++ b/src/dmd/dsymbolsem.d
@@ -4354,8 +4354,8 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
              * during static construction.
              */
             auto v = new VarDeclaration(Loc.initial, Type.tint32, Id.gate, null);
-            v.storage_class = STC.temp |
-                (scd.isSharedStaticCtorDeclaration() ? STC.shared_ : STC.static_);
+            v.storage_class = STC.temp | STC.static_ |
+                (scd.isSharedStaticCtorDeclaration() ? STC.shared_ : 0);
 
             auto sa = new Statements();
             Statement s = new ExpStatement(Loc.initial, v);
@@ -4432,8 +4432,8 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
              * during static destruction.
              */
             auto v = new VarDeclaration(Loc.initial, Type.tint32, Id.gate, null);
-            v.storage_class = STC.temp |
-                (sdd.isSharedStaticDtorDeclaration() ? STC.shared_ : STC.static_);
+            v.storage_class = STC.temp | STC.static_ |
+                (sdd.isSharedStaticDtorDeclaration() ? STC.shared_ : 0);
 
             auto sa = new Statements();
             Statement s = new ExpStatement(Loc.initial, v);
```

Allthough it seems to pass without the second line (which is weird if I understand this code correctly):
```
(scd.isSharedStaticCtorDeclaration() ? STC.shared_ : 0);
```

File: src/dmd/astbase.d [hunk 1]
```diff
@@ -109,7 +109,7 @@ struct ASTBase
         nodtor              = (1L << 24),   // don't run destructor
         nothrow_            = (1L << 25),   // never throws exceptions
         pure_               = (1L << 26),   // pure function
-        tls                 = (1L << 27),   // thread local
+        //                    (1L << 27),
```
Comment (MoonlightSentinel) [inline src/dmd/astbase.d:112]: Remove the other declaration as well:

https://github.com/dlang/dmd/blob/fe263f5991ea571b0384c6facfe207a7b17efb44/src/dmd/declaration.d#L229

**Comment by 12345swordy** [comment]
What is the hold up on here?

**Comment by Geod24** [comment]
I originally assumed it was a trivial exercise and was proven wrong. So this PR turned into a high effort / low reward deal, which is why it was stuck for so long. Came back to it recently and hopefully the first commit is the right fix, let's see what the CI says.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**